pkg/distributor/forwarding/forwarding.go

```diff
@@ -96,185 +74,260 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
-	timeout         time.Duration
-	propagateErrors bool
+type TimeseriesCounts struct {
+	SampleCount   int
+	ExemplarCount int
+}
 
-	requests prometheus.Counter
-	errors   *prometheus.CounterVec
-	samples  prometheus.Counter
-	latency  prometheus.Histogram
+func (t *TimeseriesCounts) count(ts mimirpb.PreallocTimeseries) {
+	t.SampleCount += len(ts.TimeSeries.Samples)
+	t.ExemplarCount += len(ts.TimeSeries.Exemplars)
 }
 
-func (r *request) Add(sample mimirpb.PreallocTimeseries) bool {
-	metric, err := extract.UnsafeMetricNameFromLabelAdapters(sample.Labels)
-	if err != nil {
-		// The only possible error is due to no metric name being defined, in which case we won't forward the sample
-		// to anywhere but we should still send it to the Ingesters.
-		return true
-	}
+// forwardingEndpointAssigner returns a func which takes a timeseries and assigns it endpoints to which it should be forwarded.
+func (f *forwarder) splitByTargets(tsSlice []mimirpb.PreallocTimeseries, rules validation.ForwardingRules) (TimeseriesCounts, []mimirpb.PreallocTimeseries, tsByTargets) {
+	// notIngestedCounts keeps track of the number of samples and exemplars that we don't send to the ingesters,
+	// we need to count these in order to later update some of the distributor's metrics correctly.
+	var notIngestedCounts TimeseriesCounts
+
+	// tsSliceWriteIdx is the index in the toIngest slice where we are writing TimeSeries to.
+	tsSliceWriteIdx := 0
+
+	tsByTargets := f.pools.getTsByTargets()
+	for tsSliceReadIdx, ts := range tsSlice {
+		ingest := false
+		var forwardingTarget string
+
+		metric, err := extract.UnsafeMetricNameFromLabelAdapters(ts.Labels)
+		if err != nil {
+			// Can't check whether a timeseries should be forwarded if it has no metric name.
+			// Send it to the Ingesters and don't forward it.
+			ingest = true
+		} else {
+			rule, ok := rules[metric]
+			if !ok {
+				// There is no forwarding rule for this metric, send it to the Ingesters and don't forward it.
+				ingest = true
+			} else {
+				forwardingTarget = rule.Endpoint
+				ingest = rule.Ingest
+			}
+		}
 
-	rule, ok := r.rules[metric]
-	if !ok {
-		// There is no forwarding rule for this metric, send it to the Ingesters.
-		return true
-	}
-	r.samples.Add(float64(len(sample.Samples)))
+		if forwardingTarget != "" {
+			tsByTarget, ok := tsByTargets[forwardingTarget]
+			if !ok {
+				tsByTarget.ts = f.pools.getTsSlice()
+			}
 
-	ts, ok := r.tsByEndpoint[rule.Endpoint]
-	if !ok {
-		ts = r.pools.timeseries.Get().(*[]mimirpb.PreallocTimeseries)
-		r.requests.Inc()
-	}
+			tsByTarget.ts = f.growTimeseriesSlice(tsByTarget.ts)
+			tsWriteIdx := len(tsByTarget.ts) - 1
+			mimirpb.DeepCopyTimeseries(tsByTarget.ts[tsWriteIdx].TimeSeries, ts.TimeSeries)
+			tsByTarget.counts.count(tsByTarget.ts[tsWriteIdx])
 
-	*ts = append(*ts, sample)
-	r.tsByEndpoint[rule.Endpoint] = ts
+			tsByTargets[forwardingTarget] = tsByTarget
+		}
 
-	return rule.Ingest
-}
+		if ingest {
```

Comment:
I'm assuming that we do this work managing indexes for performance reasons? Are there any numbers around the allocations this saves vs the naive case? It seems rather complex to avoid allocating an extra slice of series to ingest.

Reply:
You are correct that this is for performance reasons. I simplified the logic a bit and also commented the relevant numbers here:

https://github.com/grafana/mimir/pull/2295/#discussion_r931610962